rustls/src/compress.rs

```diff
@@ -0,0 +1,77 @@
+//! Certificate compression and decompression support
+
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::enums::CertificateCompressionAlgorithm;
+
+/// An available certificate decompression algorithm.
+pub trait CertDecompressor: Debug + Send + Sync {
+    /// Which algorithm this decompressor handles.
+    fn algorithm(&self) -> CertificateCompressionAlgorithm;
+
+    /// Decompress `input`, writing the result to `output`.
+    ///
+    /// `output` matches the declared length of the decompressed data.  The
+    /// decompression process must precisely fill `output`.  If it produces
+    /// more or fewer bytes than fit in `output`, `Err(DecompressionFailed)` must
+    /// be returned.
+    ///
+    /// `Err(DecompressionFailed)` should be returned if the `input` is in any
+    /// way malformed.
+    fn decompress(&self, input: &[u8], output: &mut [u8]) -> Result<(), DecompressionFailed>;
+}
+
+/// Returns the supported `CertDecompressor` implementations enabled
+/// by crate features.
+pub fn default_cert_decompressors() -> &'static [&'static dyn CertDecompressor] {
+    &[]
+}
+
+/// An available certificate compression algorithm.
+pub trait CertCompressor: Debug + Send + Sync {
+    /// Which algorithm this compressor handles.
+    fn algorithm(&self) -> CertificateCompressionAlgorithm;
+
+    /// Compress `input`, returning the result.
+    ///
+    /// `input` is consumed by this function so (if the underlying implementation
+    /// supports it) the compression can be performed in-place.
+    ///
+    /// `level` is a hint as to how much effort to expend on the compression.
+    ///
+    /// `Err(CompressionFailed)` may be returned for any reason.
+    fn compress(
+        &self,
+        input: Vec<u8>,
+        level: CompressionLevel,
+    ) -> Result<Vec<u8>, CompressionFailed>;
+}
+
+/// A hint for how many resources to dedicate to a compression.
+#[derive(Debug, Copy, Clone, PartialEq)]
```

Comment:
This feels like a case where deriving `Eq` is also appropriate. Am I missing a reason to avoid that?